src/Storages/ExternalStream/Kafka/Kafka.h

```diff
@@ -5,21 +5,45 @@
 #include <Storages/ExternalStream/ExternalStreamSettings.h>
 #include <Storages/ExternalStream/StorageExternalStreamImpl.h>
 #include <Storages/ExternalStream/ExternalStreamCounter.h>
+#include <Storages/ExternalStream/Kafka/Consumer.h>
+#include <Storages/ExternalStream/Kafka/ConsumerPool.h>
+#include <Storages/ExternalStream/Kafka/Producer.h>
 #include <Storages/Streaming/SeekToInfo.h>
 
 namespace DB
 {
 
 class IStorage;
 
+
 class Kafka final : public StorageExternalStreamImpl
 {
 public:
+    using ConfPtr = std::unique_ptr<rd_kafka_conf_t, decltype(rd_kafka_conf_destroy) *>;
+
+    static Poco::Logger * cbLogger() {
+        static Poco::Logger * logger { &Poco::Logger::get("KafkaExternalStream") };
+        return logger;
+    }
+
+    static int onStats(struct rd_kafka_s * rk, char * json, size_t json_len, void * opaque);
+    static void onError(struct rd_kafka_s * rk, int err, const char * reason, void * opaque);
+    static void onThrottle(struct rd_kafka_s * rk, const char * broker_name, int32_t broker_id, int throttle_time_ms, void * opaque);
+
     Kafka(IStorage * storage, std::unique_ptr<ExternalStreamSettings> settings_, const ASTs & engine_args_, bool attach, ExternalStreamCounterPtr external_stream_counter_, ContextPtr context);
     ~Kafka() override = default;
 
-    void startup() override { }
-    void shutdown() override { }
+    void startup() override { LOG_INFO(logger, "Starting Kafka External Stream"); }
+    void shutdown() override {
+        LOG_INFO(logger, "Shutting down Kafka External Stream");
+        /// Must release all resources here rather than relying on the deconstructor.
+        /// Because the `Kafka` instance will not be destroyed immediately when the external stream gets dropped.
+        consumer_pool.reset();
+        if (producer_topic)
+            producer_topic.reset();
+        if (producer)
+            producer.reset();
```

Comment:
Is the release order required ? 
1) release producer 
2) release consumer
3) release topic

Reply:
Yes, it should release the `producer_topic` before releasing the `producer`, because the `producer_topic` contains a reference to the `producer`.

Reply:
In my understanding, each producer and consumer has its own Topic, we can move the ownership of Topic to Producer and Consumer

Reply:
In my understanding, each producer and consumer has its own Topic, we can move the ownership of Topic to Producer and Consumer

Reply:
Technically speaking, we can make a Producer/Consumer owns the Topic, but in this case, each `Kafka` has its own Consumers, Producer and Topic, it owns everything. So, for now, I just want to keep things simple, and each class just represents the same concept in the librdkafka library. Later on, when we work on the `Connection` concept, and start reusing connections between external streams, we will revisit these concepts and figure out how to do it properly.